src/azure-cli/azure/cli/command_modules/role/custom.py

```diff
@@ -853,13 +853,17 @@ def create_application(cmd, display_name, homepage=None, identifier_uris=None,
 
 def _get_grant_permissions(graph_client, client_sp_object_id=None, query_filter=None):
     query_filter = query_filter or ("clientId eq '{}'".format(client_sp_object_id) if client_sp_object_id else None)
+    grant_info = graph_client.oauth2_permission_grant.list(filter=query_filter)
     try:
-        grant_info = graph_client.oauth2_permission_grant.list(filter=query_filter)
-    except CloudError as ex:  # Graph doesn't follow the ARM error; otherwise would be caught by msrest-azure
+        # Make the REST request immediately so that errors can be raised and handled.
+        return list(grant_info)
+    except CloudError as ex:
         if ex.status_code == 404:
-            return []
+            raise CLIError("Service principal with appId or objectId '{id}' doesn't exist. "
+                           "If '{id}' is an appId, make sure an associated service principal is created "
+                           "for the app. To create one, run `az ad sp create --id {id}`."
+                           .format(id=client_sp_object_id))
```

Comment:
Shall we use a UserFault type error here?

Reply:
`InvalidArgumentValueError` for example

Reply:
This is rather complex here as `client_sp_object_id` is indeed a valid `appId` but the app doesn't have associated SP. This command group is already considered deprecated (implicitly) anyway.